File: ios/Reguerta/Reguerta/Data/Notifications/InMemoryNotificationRepository.swift
import Foundation

actor InMemoryNotificationRepository: NotificationRepository {
    private var notifications: [String: NotificationEvent]
    private var readNotificationIdsByMember: [String: Set<String>]

    init(items: [NotificationEvent]? = nil) {
        self.init(items: items, readNotificationIdsByMember: [:])
    }

    init(
        items: [NotificationEvent]? = nil,
        readNotificationIdsByMember: [String: Set<String>]
    ) {
        self.notifications = Dictionary(uniqueKeysWithValues: (items ?? Self.seedNotifications).map { ($0.id, $0) })
        self.readNotificationIdsByMember = readNotificationIdsByMember
    }

    private static let seedNotifications: [NotificationEvent] = [
        NotificationEvent(
            id: "notification_welcome_001",
            title: "seed.notification.welcome.title",
            body: "seed.notification.welcome.body",
            type: "admin_broadcast",
            target: "all",
            userIds: [],
            segmentType: nil,
            targetRole: nil,
            createdBy: "system",
            sentAtMillis: 1_711_849_600_000,
            weekKey: nil
        ),
        NotificationEvent(
            id: "notification_admin_001",
            title: "seed.notification.admin_channel.title",
            body: "seed.notification.admin_channel.body",
            type: "admin_broadcast",
            target: "segment",
            userIds: [],
            segmentType: "role",
            targetRole: .admin,
            createdBy: "system",
            sentAtMillis: 1_712_108_800_000,
            weekKey: nil
        )
    ]

    func allNotifications() async -> [NotificationEvent] {
        var localized: [NotificationEvent] = []
        localized.reserveCapacity(notifications.count)
        for event in notifications.values {
            localized.append(await localizedSeedNotification(event))
        }
        return localized.sorted { $0.sentAtMillis > $1.sentAtMillis }
    }

    func readNotificationIds(memberId: String) async -> Set<String> {
        readNotificationIdsByMember[memberId] ?? []
    }

    func markNotificationsRead(memberId: String, notificationIds: [String], readAtMillis _: Int64) async {
        guard !notificationIds.isEmpty else { return }
        var readIds = readNotificationIdsByMember[memberId] ?? []
        readIds.formUnion(notificationIds)
        readNotificationIdsByMember[memberId] = readIds
    }

    func send(event: NotificationEvent) async -> NotificationEvent {
        let eventId = event.id.isEmpty ? "notification_\(notifications.count + 1)" : event.id
        let persisted = NotificationEvent(
            id: eventId,
            title: event.title,
            body: event.body,
            type: event.type,
            target: event.target,
            userIds: event.userIds,
            segmentType: event.segmentType,
            targetRole: event.targetRole,
            createdBy: event.createdBy,
            sentAtMillis: event.sentAtMillis,
            weekKey: event.weekKey
        )
        notifications[eventId] = persisted
        return persisted
    }
}

private func localizedSeedNotification(_ event: NotificationEvent) async -> NotificationEvent {
    let title = await MainActor.run {
        NSLocalizedString(event.title, comment: "")
    }
    let body = await MainActor.run {
        NSLocalizedString(event.body, comment: "")
    }
    return NotificationEvent(
        id: event.id,
        title: title,
        body: body,
        type: event.type,
        target: event.target,
        userIds: event.userIds,
        segmentType: event.segmentType,
        targetRole: event.targetRole,
        createdBy: event.createdBy,
        sentAtMillis: event.sentAtMillis,
        weekKey: event.weekKey
    )
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/data/notifications/InMemoryNotificationRepository.kt
package com.reguerta.user.data.notifications

import com.reguerta.user.domain.access.MemberRole
import com.reguerta.user.domain.notifications.NotificationEvent
import com.reguerta.user.domain.notifications.NotificationRepository
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class InMemoryNotificationRepository : NotificationRepository {
    private val mutex = Mutex()
    private val notifications = mutableMapOf(
        "notification_welcome_001" to NotificationEvent(
            id = "notification_welcome_001",
            title = "Bienvenida a La Reguerta",
            body = "Aquí verás recordatorios y avisos extraordinarios enviados por la administración.",
            type = "admin_broadcast",
            target = "all",
            userIds = emptyList(),
            segmentType = null,
            targetRole = null,
            createdBy = "system",
            sentAtMillis = 1_711_849_600_000,
            weekKey = null,
        ),
        "notification_admin_001" to NotificationEvent(
            id = "notification_admin_001",
            title = "Canal admin disponible",
            body = "Este entorno de pruebas ya puede enviar notificaciones extraordinarias desde la app.",
            type = "admin_broadcast",
            target = "segment",
            userIds = emptyList(),
            segmentType = "role",
            targetRole = MemberRole.ADMIN,
            createdBy = "system",
            sentAtMillis = 1_712_108_800_000,
            weekKey = null,
        ),
    )
    private val readNotificationIdsByMember = mutableMapOf<String, MutableSet<String>>()

    override suspend fun getAllNotifications(): List<NotificationEvent> = mutex.withLock {
        notifications.values.sortedByDescending { it.sentAtMillis }
    }

    override suspend fun getReadNotificationIds(memberId: String): Set<String> = mutex.withLock {
        readNotificationIdsByMember[memberId]?.toSet().orEmpty()
    }

    override suspend fun markNotificationsRead(
        memberId: String,
        notificationIds: Set<String>,
        readAtMillis: Long,
    ) = mutex.withLock {
        if (notificationIds.isEmpty()) return@withLock
        val readIds = readNotificationIdsByMember.getOrPut(memberId) { mutableSetOf() }
        readIds.addAll(notificationIds.filter(String::isNotBlank))
    }

    override suspend fun sendNotification(event: NotificationEvent): NotificationEvent = mutex.withLock {
        val eventId = event.id.ifBlank { "notification_${notifications.size + 1}" }
        val persisted = event.copy(id = eventId)
        notifications[eventId] = persisted
        persisted
    }
}
